### PR TITLE
PM-9681: Setup Bitwarden PIN

### PR DESCRIPTION
**Pull request by shannon-livefront**
## 🎟️ Tracking

[PM-9681](https://bitwarden.atlassian.net/browse/PM-9681)

## 📔 Objective

If the user is trying to save or select a passkey with their vault unlocked, a device with no lock screen, and no PIN or password, prompt them to set up a Bitwarden PIN before saving or selecting the passkey on the item listing view.

## 📸 Screenshots

| Before | After |
| --- | --- |
| <video src="https://github.com/user-attachments/assets/8bfd3cfd-59e0-43bc-90be-8e8878aaf30c" /> | <video src="https://github.com/user-attachments/assets/76ef5363-07da-4a0b-a649-1302ece700b6" /> |

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[PM-9681]: https://bitwarden.atlassian.net/browse/PM-9681?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingScreen.kt [hunk 1]
```diff
@@ -329,6 +349,31 @@ private fun VaultItemListingDialogs(
             )
         }
 
+        is VaultItemListingState.DialogState.Fido2PinSetUpPrompt -> {
+            var pin by rememberSaveable { mutableStateOf("") }
+            PinInputDialog(
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingScreen.kt:354]: Would it make sense to make a helper function that lives with the existing `PinInputDialog`

```kotlin
/**
 * Documentation!!!
 */
@Composable
fun PinInputDialog(
    onCancelClick: () -> Unit,
    onSubmitClick: (String) -> Unit,
    onDismissRequest: () -> Unit,
) {
    var pin by remember { mutableStateOf("") }
    PinInputDialog(
        pin = pin,
        onPinChange = { pin = it },
        onCancelClick = onCancelClick,
        onSubmitClick = { onSubmitClick(pin) },
        onDismissRequest = onDismissRequest,
    )
}
```

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingScreen.kt [hunk 2]
```diff
@@ -329,6 +349,31 @@ private fun VaultItemListingDialogs(
             )
         }
 
+        is VaultItemListingState.DialogState.Fido2PinSetUpPrompt -> {
+            var pin by rememberSaveable { mutableStateOf("") }
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingScreen.kt:353]: I thought we decided that pins should not be saved?

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingScreen.kt (hunk 1 above)
Comment (shannon-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingScreen.kt:354]: That's clever, thanks!

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingScreen.kt (hunk 2 above)
Comment (shannon-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingScreen.kt:353]: Yeah, you're right - I was trying to make it the same as the `AccountSecurityScreen`, but you're right that it should be updated there as well!

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingViewModel.kt [hunk 3]
```diff
@@ -426,6 +439,44 @@ class VaultItemListingViewModel @Inject constructor(
...
+                        selectedCipherId = action.selectedCipherId,
+                    ),
+                )
+            }
+            return
+        }
+
+        // There's no need to ask the user whether or not they want to use their master password
+        // on login, and shouldRequireMasterPasswordOnRestart is hardcoded to false, because the
+        // user can only reach this part of the flow if they have no master password.
```
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/itemlisting/VaultItemListingViewModel.kt:458]: 👍

**Review by SaintPatrck — APPROVED** [review]

**Pull request merged**